process/block/preprocess/transactions.go

```diff
@@ -471,9 +473,175 @@ func (txs *transactions) CreateAndProcessMiniBlocks(
+		sameType := miniBlock.Type == mergedMiniBlock.Type
+		sameSenderShard := miniBlock.SenderShardID == mergedMiniBlock.SenderShardID
+		sameReceiverShard := miniBlock.ReceiverShardID == mergedMiniBlock.ReceiverShardID
+
+		canMerge := sameSenderShard && sameReceiverShard && sameType
+		if canMerge {
+			mergedMiniBlock.TxHashes = append(mergedMiniBlock.TxHashes, miniBlock.TxHashes...)
+			return mergedMiniBlocks
+		}
+	}
+
+	mergedMiniBlocks = append(mergedMiniBlocks, miniBlock)
+
+	return mergedMiniBlocks
+}
+
+func (txs *transactions) expandMiniBlocks(miniBlocks block.MiniBlockSlice) block.MiniBlockSlice {
+	if len(miniBlocks) == 0 {
+		return miniBlocks
+	}
+
+	mbsToExpand := make(block.MiniBlockSlice, 0)
+
+	for i := 0; i < len(miniBlocks); i++ {
+		if miniBlocks[i].SenderShardID == txs.shardCoordinator.SelfId() {
+			mbsToExpand = append(mbsToExpand, miniBlocks[i])
+		}
+	}
+
+	if len(mbsToExpand) > 0 {
+		expandedMbs := txs.expand(mbsToExpand)
+		if len(expandedMbs) > 0 {
+			return expandedMbs
+		}
+	}
+
+	return miniBlocks
+
+}
+
+func (txs *transactions) expand(miniBlocks block.MiniBlockSlice) block.MiniBlockSlice {
+	if len(miniBlocks) == 0 {
+		return miniBlocks
+	}
+
+	mbSlice := make(block.MiniBlockSlice, 0)
+	txsInfo := make([][]*txInfo, txs.shardCoordinator.NumberOfShards())
+	mapSenderNonce := make(map[string]uint64)
+
+	for i := 0; i < len(miniBlocks); i++ {
+		miniBlock := miniBlocks[i]
+		txsInfo[miniBlock.ReceiverShardID] = make([]*txInfo, 0)
+		for _, txHash := range miniBlocks[i].TxHashes {
+			txNfo, ok := txs.txsForCurrBlock.txHashAndInfo[string(txHash)]
+			if !ok {
+				continue
+			}
+			txNfo.txHash = txHash
+			txsInfo[miniBlock.ReceiverShardID] = append(txsInfo[miniBlock.ReceiverShardID], txNfo)
+		}
+	}
+
+	for {
+		foundSomething := false
+		mbSlice, foundSomething = txs.tryCreateMiniBlocks(txsInfo, mapSenderNonce, mbSlice, foundSomething)
+		if !foundSomething {
+			break
+		}
+	}
+
+	return mbSlice
+}
+
+func (txs *transactions) tryCreateMiniBlocks(
+	txsInfo [][]*txInfo,
+	mapSenderNonce map[string]uint64,
+	mbSlice block.MiniBlockSlice,
+	foundSomething bool,
+) (block.MiniBlockSlice, bool) {
+
+	var mb *block.MiniBlock
+
+	for i := 0; i < len(txsInfo); i++ {
+		if len(txsInfo[i]) > 0 {
+			mb, txsInfo[i] = txs.createMiniBlockForShard(txsInfo[i], mapSenderNonce)
+			if len(mb.TxHashes) > 0 {
+				mbSlice = append(mbSlice, mb)
+				foundSomething = true
+			}
+		}
+	}
+
+	return mbSlice, foundSomething
+}
+
+func (txs *transactions) createMiniBlockForShard(
+	currentTxs []*txInfo,
+	mapSenderNonce map[string]uint64,
+) (*block.MiniBlock, []*txInfo) {
+
+	mb := &block.MiniBlock{}
+	mb.SenderShardID = txs.shardCoordinator.SelfId()
+	mb.TxHashes = make([][]byte, 0)
+	mb.Type = block.TxBlock
+
+	for {
+		found := false
+		for _, tx := range currentTxs {
+			mb.ReceiverShardID = tx.receiverShardID
+			nonce, ok := mapSenderNonce[string(tx.tx.GetSndAddress())]
+			if !ok || tx.tx.GetNonce() == nonce+1 {
```

Comment:
tx.tx does not look good. rename variables.

Reply:
Refactored